restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.web.docsapi.service;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import io.reactivex.Flowable;
+import io.reactivex.Single;
+import io.stargate.db.ImmutableParameters;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.datastore.Row;
+import io.stargate.db.query.builder.AbstractBound;
+import io.stargate.db.query.builder.BuiltSelect;
+import io.stargate.db.schema.Column;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/** Executes pre-built document queries, groups document rows and manages document pagination. */
+public class QueryExecutor {
+  private final Accumulator TERM = new Accumulator();
+
+  private final DataStore dataStore;
+
+  public QueryExecutor(DataStore dataStore) {
+    this.dataStore = dataStore;
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    return queryDocs(1, query, pageSize, pagingState);
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      int identityDepth, AbstractBound<?> query, int pageSize, ByteBuffer pagingState) {
+    BuiltSelect select = (BuiltSelect) query.source().query();
+    if (identityDepth < 1 || identityDepth > select.table().primaryKeyColumns().size()) {
+      throw new IllegalArgumentException("Invalid document identity depth: " + identityDepth);
+    }
+
+    List<Column> idColumns = select.table().primaryKeyColumns().subList(0, identityDepth);
+
+    return execute(query, pageSize, pagingState)
```

Comment:
What's the runtime complexity of this? Seems like it could be O(number_of_rows^2).

I admit to no fully understanding this, but why not accumulate into a `HashMap` by `id`? (This is greatly simplified)

```java
return execute(query, pageSize, pagingState)
        .collect(HashMap<String, Accumulator>::new, (map, rs) -> {
          for (Row row : rs.rows()) {
            map.computeIfAbsent(row.id(), Accumulator::new).rows.add(row);
          }   
        })  
        .map(HashMap::values)
        .flattenAsFlowable(a -> a)
        .map(Accumulator::toDoc);

```

Working example: https://gist.github.com/mpenick/aec03128af80e668a4676705773cc707

Reply:
Will `collect` require iterating the whole result set? The main idea of current impl. is to avoid pre-loading more rows than required and support resuming on the next call with a (custom) paging state.

Reply:
Not sure about `O(number_or_rows ^ 2)` :thinking: Each `Accumulator` adds elements to the same `List` until it is "complete". So there's some overhead, but it should not be about `O(1)`... did I miss something?

Reply:
Maybe I'm misunderstanding this. 

Doesn't `seeds()` create an accumulator for every row? https://github.com/stargate/stargate/pull/937/files#diff-d2960a88647c3594554b5beca0f8776780d83adb9b9efb9b15e519ff195a008bR113

Reply:
Right, so the total complexity should be `O(number_of_rows)`. By `O(1)` I meant overhead.